src/backend/egl/display.rs

```diff
@@ -212,6 +201,100 @@ impl EGLDisplay {
         })
     }
 
+    /// Create a new [`EGLDisplay`] from an already initialized EGLDisplay and EGLConfig handle
+    pub unsafe fn from_raw<L>(
+        display: *const c_void,
+        config: *const c_void,
+        logger: L,
+    ) -> Result<EGLDisplay, Error>
+    where
+        L: Into<Option<::slog::Logger>>,
+    {
+        let log = crate::slog_or_fallback(logger.into()).new(o!("smithay_module" => "backend_egl"));
+
+        let dp_extensions = ffi::make_sure_egl_is_loaded()?;
+        debug!(log, "Supported EGL client extensions: {:?}", dp_extensions);
+
+        let egl_version = {
+            let p = CStr::from_ptr(
+                wrap_egl_call(|| ffi::egl::QueryString(display, ffi::egl::VERSION as i32))
+                    .map_err(|_| Error::DisplayQueryResultInvalid)?,
+            );
+
+            let version_string = String::from_utf8(p.to_bytes().to_vec()).unwrap_or_else(|_| String::new());
+            let mut version_iterator = version_string
+                .split(' ')
+                .next()
+                .ok_or(Error::DisplayQueryResultInvalid)?
+                .split('.');
+
+            let major = version_iterator
+                .next()
+                .and_then(|v| v.parse::<i32>().ok())
+                .ok_or(Error::DisplayQueryResultInvalid)?;
+            let minor = version_iterator
+                .next()
+                .and_then(|v| v.parse::<i32>().ok())
+                .ok_or(Error::DisplayQueryResultInvalid)?;
+
+            info!(log, "EGL Version: {:?}", (major, minor));
+            (major, minor)
+        };
+
+        let extensions = EGLDisplay::get_extensions(egl_version, display)?;
+        info!(log, "Supported EGL display extensions: {:?}", extensions);
+
+        let (dmabuf_import_formats, dmabuf_render_formats) =
+            get_dmabuf_formats(&display, &extensions, &log).map_err(Error::DisplayCreationError)?;
+
+        let egl_api =
+            wrap_egl_call(|| ffi::egl::QueryAPI()).map_err(|_| Error::OpenGlesNotSupported(None))?;
+        if egl_api != ffi::egl::OPENGL_ES_API {
```

Comment:
@drakulix, this condition isn't strictly needed?

Reply:
On mesa? No. On other drivers? Yes.
OpenGL ES can behave subtly differently and most importantly `GL_OES_EGL_image` is a OpenGL ES only protocol.
Mesa does expose it for desktop GL contexts as well, but other drivers do not. Therefore we strictly need an ES context.

So yes, this check is necessary.